File: solutions/go/02/greedy_intervals/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

type interval struct{ start, end int }

func main() {
	in := bufio.NewReader(os.Stdin)
	var n int
	fmt.Fscan(in, &n)
	ivs := make([]interval, n)
	for i := 0; i < n; i++ {
		fmt.Fscan(in, &ivs[i].start, &ivs[i].end)
	}
	sort.Slice(ivs, func(i, j int) bool {
		return ivs[i].end < ivs[j].end
	})
	last := -1 << 60
	cnt := 0
	for _, iv := range ivs {
		if iv.start >= last {
			cnt++
			last = iv.end
		}
	}
	fmt.Println(cnt)
}

File: solutions/python/02/greedy_intervals/main.py
import sys


def main() -> None:
    input = sys.stdin.readline
    n = int(input())
    intervals = []
    for _ in range(n):
        l, r = map(int, input().split())
        intervals.append((l, r))
    intervals.sort(key=lambda x: x[1])

    last = -10**18
    cnt = 0
    for l, r in intervals:
        if l >= last:
            cnt += 1
            last = r
    print(cnt)


if __name__ == "__main__":
    main()
